src/backend/distributed/utils/metadata_cache.c

```diff
@@ -933,6 +934,52 @@ HasUninitializedShardInterval(ShardInterval **sortedShardIntervalArray, int shar
 
 
 /*
+ * HasOverlappingShardInterval determines whether the given list of sorted
+ * shards has overlapping ranges.
+ */
+static bool
+HasOverlappingShardInterval(ShardInterval **shardIntervalArray,
```

Comment:
I think the main distinction between range- and append-partitioning is that ranges are distinct. It would be better to enforce this during shard creation. I'd prefer if we can trim down the code a bit and also lower the cost of cache invalidations with high shard counts.